man/promote_content_views.8
.\" Manpage for sat_import.
.\" Contact ggatward@redhat.com to correct errors or typos.
.TH PROMOTE_CONTENT_VIEWS 8 "04 Jan 2017" "sat6_scripts" "sat6_scripts User Manual" man page"
.SH NAME
promote_content_views \- promote content views defined within Satellite 6

.SH SYNOPSIS
.B promote_content_views [\-o ORGANISATION] [\-e ENVIRONMENT] [\-a] [\-d] [\-l]
.LP
.B "promote_content_views --help"

.SH DESCRIPTION
.B promote_content_views
is a custom python script for promoting defined content views within a Satellite 6 server through the defined lifecycle environments.
The target environment is specified as a parameter, and content views from the previous environment will be promoted to the given target.
.PP
For example, given a lifecycle of
.B "Library -> Quality -> Production"
, specifying Production as the target will promote content views from Quality to Production environments.


.SH OPTIONS
The options that apply to the
.B promote_content_views
command are:
.PP
.BR "-o", " --org"
.I "ORGANISATION"
.RS 3
The organisation within Satellite to perform the activity for. The default organisation is defined in the
.I config.yml
file, but can be overridden with this option.
.RE
.PP
.BR "-e", " --env"
.I "ENVIRONMENT"
.RS 3
The target environment to promote to. The environment defined must exist in the Satellite.
Optionally, environments can also be specified in the
.I config.yml
file, defining specific content views to include in the promotion to the given environment - see
.BR sat6_scripts (8).
.RE
.PP
.BR "-a", " --all"
.RS 3
Promote all configured content views. By default
.B promote_content_views
will publish only the content views defined in the
.I config.yml
file for the nominated environment, if an environment definition exists in the config.yml file.
If no environment definition is found, all content views will be promoted from the previous environment.
.RE
.PP
.BR "-l", " --last"
.RS 3
Display the dates of the last successful publish and promotions.
.RE
.PP
.BR "-d", " --dryrun"
.RS 3
Do not perform the actual promotion activity, but show what would be done.
This option can be used to verify that the correct configuration options have been defined and applied before running the actual promotion of content.
.RE
.PP
.BR "-q", " --quiet"
.RS 3
Do not display refreshing progress bars during the operation.
This option can be used when including this script in automated tasks, for example from cron.
.RE
.PP
.BR "-m", " --forcemeta"
.RS 3
Force the regeneration of repository metadata during the publish.
.RE


.SH EXAMPLES
Check when the last promotes were performed:
.RS 6
.B "promote_content_views -l"
.RE

Promote content views to the Quality environment, using the definition in the config.yml:
.RS 6
.B "promote_content_views -e Quality"
.RE

Show what will be promoted by default to the Production environment, but do not actually promote the content:
.RS 6
.B "promote_content_views -e Production -d"
.RE

.SH FILES
.B Main Configuration
.RS 3
.I /usr/share/sat6_scripts/config/config.yml
.RE
.LP
.B Log Files
.RS 3
.I /var/log/satellite/*
.RE

.SH SEE ALSO
.BR sat6_scripts_config (8),
.BR publish_content_views (8),
.BR clean_content_views (8).

.SH AUTHOR
Geoff Gatward <ggatward@redhat.com>


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]


## [1.2.2] - 2018-02-25
### Added
- Option to tag a published content view with a custom comment
- Option to clean versions taking into account published versions
- Option to force metadata regeneration during publish and promote

### Fixed
- Fix unreferenced var in check_disk_space function
- Various minor README corrections
- auto_content did not publish CV if the satellite is internet connected
- publish_content_views KeyError exception if API returns null value
- Fix handling empty lists for used and unused versions in clean_content_views


## [1.2.1] - 2017-12-11
### Fixed
- Incorrect mapping of days in auto_content.py
- Support for username:token authentication to Artifactory Puppet Forge server
- Email now supports multiple recipients


## [1.2.0] - 2017-12-10
### Added
- push_puppetforge now supports jFrog Artifiactory repository via HTTP POST
- sat_import now checks for exports that have not been imported (missed/skipped)
- sat_import --fixhistory option to force align import/export histories
- Email notification capability for use when automating content scripts
- Add unattended option to allow scripts to be automated
- auto_content scripts to allow unattended import/publish/promote/clean activity

### Changed
- --notar export saved in /cdn_export dir rather than /export to prevent it being deleted

### Removed
- Skip GPG short option (-n)


## [1.1.1] - 2017-10-25
### Added
- Allow limiting of Publish and Promotion to specified number of CV's at once.


## [1.1.0] - 2017-10-23
### Added
- Proxy support for download_manifest
- Quiet option in publish/promote scripts
- Full repo package comparison written to log during disconnected import
- Provide a history of successful content exports and imports
- Present a warning if import has already been performed for given dataset
- Option to leave export in unarchived format (--notar)
- Option to force a disconnected (import only) Satellite to export if required

### Changed
- Updated exit codes throughout the scripts
- Updated export dataset name format to allow for multiple exports per day
- Updated to Semantic Versioning

### Fixed
- Various minor bugfixes and typos
- Fixed bug in DoV export

### Deprecated
- Skip GPG short option (-n) will be removed in 1.2.0. Long option (--nogpg) will remain.


## [1.0] - 2017-03-06
- Initial 'public' release

## 0.6 - 2017-02-27
- Last of a series of pre-release betas

[Unreleased]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.2...HEAD
[1.2.2]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.1...1.2.2
[1.2.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/1.0...1.1.0
[1.0]: https://github.com/RedHatSatellite/sat6_scripts/compare/0.6...1.0


promote_content_views.py
#!/usr/bin/python
#title           :promote_content_view.py
#description     :Promotes Satellite 6 content view versions
#URL             :https://github.com/ggatward/sat6_scripts
#author          :Geoff Gatward <ggatward@redhat.com>
#notes           :This script is NOT SUPPORTED by Red Hat Global Support Services.
#license         :GPLv3
#==============================================================================
"""
Promotes Content Views from the previous lifecycle environment.

"""
#pylint: disable-msg=R0912,R0913,R0914,R0915

import sys
import os
import argparse
import datetime
import pickle
import simplejson as json
import helpers

try:
    import yaml
except ImportError:
    print "Please install the PyYAML module."
    sys.exit(1)


# Get the details about the environments
def get_envs(org_id):
    """Get list of environments for the given org"""
    envs = helpers.get_json(
        helpers.SAT_API + "organizations/" + str(org_id) + "/environments/")

    # ... and add them to a dictionary, with respective 'Prior' environment
    env_list = {}
    prior_list = {}
    for env in envs['results']:
        env_list[env['name']] = env['id']
        if env['name'] == "Library":
            prior = 0
        else:
            prior = env['prior']['id']
        prior_list[env['id']] = prior

        msg = "Found environment '" + env['name'] + "', env_id " + str(env['id']) +\
            " (prior_id " + str(prior) + ")"
        helpers.log_msg(msg, 'DEBUG')

    return env_list, prior_list


# Get details about Content Views and versions
def get_cv(org_id, target_env, env_list, prior_list, promote_list):
    """Get the content views"""
    # Find the ID of the environment we are promoting to and from
    if not target_env in env_list:
        msg = "Target environment '" + target_env + "' not found"
        helpers.log_msg(msg, 'ERROR')
        if helpers.MAILOUT:
            helpers.tf.seek(0)
            output = "{}".format(helpers.tf.read())
            helpers.mailout(helpers.MAILSUBJ_FP, output)
        sys.exit(1)
    else:
        target_env_id = env_list[target_env]
        source_env_id = prior_list[target_env_id]

    # Query API to get all content views for our org
    cvs = helpers.get_json(
        helpers.KATELLO_API + "organizations/" + str(org_id) + "/content_views/")
    ver_list = {}
    ver_descr = {}
    ver_version = {}

    for cv_result in cvs['results']:
        # We will never promote to/from the DOV
        if cv_result['name'] != "Default Organization View":

            # Handle specific includes and excludes
            if promote_list and cv_result['name'] not in promote_list:
                msg = "Skipping content view '" + cv_result['name'] + "'"
                helpers.log_msg(msg, 'DEBUG')
                continue

            # Get the ID of each Content View
            msg = "Processing content view '" + cv_result['name'] + "'"
            helpers.log_msg(msg, 'DEBUG')

            # Find the current version of the view in the env we are coming from
            for ver in cv_result['versions']:
                msg = "  Found in env_id " + str(ver['environment_ids']) + " view_id " +\
                    str(ver['id'])
                helpers.log_msg(msg, 'DEBUG')

                if source_env_id in ver['environment_ids']:
                    # Extract the name of the source environment so we can inform the user
                    for key, val in env_list.items():
                        if val == source_env_id:
                            prior_env = key
                    msg = "Found promotable version " + ver['version'] + " of '" +\
                        cv_result['name'] + "' in " + prior_env
                    helpers.log_msg(msg, 'INFO')
                    print msg

                    # Create a dictionary of CV IDs and the CV vers ID to promote
                    ver_list[cv_result['id']] = ver['id']
                    ver_descr[cv_result['id']] = cv_result['name']
                    ver_version[cv_result['id']] = ver['version']

    return ver_list, ver_descr, ver_version


# Promote a content view version
def promote(target_env, ver_list, ver_descr, ver_version, env_list, prior_list, dry_run, quiet, forcemeta):
    """Promote Content View"""
    target_env_id = env_list[target_env]
    source_env_id = prior_list[target_env_id]

    # Extract the name of the source environment so we can inform the user
    for key, val in env_list.items():
        if val == source_env_id:
            prior_env = key

    # Set the task name to be displayed in the task monitoring stage
    task_name = "Promotion from " + prior_env + " to " + target_env

    # Now we have all the info needed, we can actually trigger the promotion.
    # Loop through each CV with promotable versions
    task_list = []
    ref_list = {}

    # Catch scenario that no CV versions are found matching promotion criteria
    if not ver_list:
        msg = "No content view versions found matching promotion criteria"
        helpers.log_msg(msg, 'WARNING')
        if helpers.MAILOUT:
            helpers.tf.seek(0)
            output = "{}".format(helpers.tf.read())
            helpers.mailout(helpers.MAILSUBJ_FP, output)
        sys.exit(1)

    # Break repos to promote into batches as configured in config.yml
    cvchunks = [ ver_list.keys()[i:i+helpers.PROMOTEBATCH] for i in range(0, len(ver_list), helpers.PROMOTEBATCH) ]

    # Loop through the smaller subsets of repo id's
    for chunk in cvchunks:
        for cvid in chunk:

            # Check if there is a publish/promote already running on this content view
            locked = helpers.check_running_publish(cvid, ver_descr[cvid])

            if not locked:
                msg = "Promoting '" + str(ver_descr[cvid]) + "' Version " + str(ver_version[cvid]) +\
                    " from " + prior_env + " to " + str(target_env)
                helpers.log_msg(msg, 'INFO')
                print helpers.HEADER + msg + helpers.ENDC

            if not dry_run and not locked:
                try:
                    task_id = helpers.post_json(
                        helpers.KATELLO_API + "content_view_versions/" + str(ver_list[cvid]) +\
                        "/promote/", json.dumps(
                            {
                                "environment_id": target_env_id,
                                "force_yum_metadata_regeneration": str(forcemeta)
                            }
                            ))["id"]
                except Warning:
                    msg = "Failed to initiate promotion of " + str(ver_descr[cvid])
                    helpers.log_msg(msg, 'WARNING')
                else:
                    task_list.append(task_id)
                    ref_list[task_id] = ver_descr[cvid]

        # Exit in the case of a dry-run
        if dry_run:
            msg = "Dry run - not actually performing promotion"
            helpers.log_msg(msg, 'WARNING')
        else:
            # Monitor the status of the promotion tasks
            helpers.watch_tasks(task_list, ref_list, task_name, quiet)

    # Exit in the case of a dry-run
    if dry_run:
        sys.exit(2)
    else:
        return


def main(args):
    """
    Main routine
    """

    # Who is running this script?
    runuser = helpers.who_is_running()

    # Set the base dir of the script and where the var data is
    global dir
    global vardir
    dir = os.path.dirname(__file__)
    vardir = os.path.join(dir, 'var')

    # Check for sane input
    parser = argparse.ArgumentParser(
        description='Promotes content views for specified organization to the target environment.')
    group = parser.add_mutually_exclusive_group()
    # pylint: disable=bad-continuation
    parser.add_argument('-e', '--env', help='Target Environment (e.g. Development, Quality, Production)',
        required=False)
    parser.add_argument('-o', '--org', help='Organization (Uses default if not specified)',
        required=False)
    group.add_argument('-a', '--all', help='Promote ALL content views', required=False,
        action="store_true")
    parser.add_argument('-d', '--dryrun', help='Dry Run - Only show what will be promoted',
        required=False, action="store_true")
    parser.add_argument('-l', '--last', help='Display last promotions', required=False,
        action="store_true")
    parser.add_argument('-q', '--quiet', help="Suppress progress output updates", required=False,
        action="store_true")
    parser.add_argument('-m', '--forcemeta', help="Force metadata regeneration", required=False,
        action="store_true")

    args = parser.parse_args()

    # Log the fact we are starting
    msg = "-------- Content view promotion started by " + runuser + " -----------"
    helpers.log_msg(msg, 'INFO')

    # Set our script variables from the input args
    if args.org:
        org_name = args.org
    else:
       org_name = helpers.ORG_NAME
    target_env = args.env
    dry_run = args.dryrun

    # Load the promotion history
    if not os.path.exists(vardir + '/promotions.pkl'):
        if not os.path.exists(vardir):
            os.makedirs(vardir)
        phistory = {}
    else:
        phistory = pickle.load(open(vardir + '/promotions.pkl', 'rb'))

    # Read the promotion history if --last requested
    if args.last:
        if phistory:
            print 'Last promotions:'
            for lenv, time in phistory.iteritems():
                print lenv, time
        else:
            print 'No promotions recorded'
        sys.exit(0)

    # Error if no environment to promote to is given
    if args.env is None:
        parser.error('--env is required')

    promote_list = []
    if not args.all:
        for x in helpers.CONFIG['promotion']:
            if x == 'batch':
                continue
            if helpers.CONFIG['promotion'][x]['name'] == target_env:
                promote_list = helpers.CONFIG['promotion'][x]['content_views']

        if not promote_list:
            msg = "Cannot find promotion configuration for '" + target_env + "'"
            helpers.log_msg(msg, 'ERROR')
            if helpers.MAILOUT:
                helpers.tf.seek(0)
                output = "{}".format(helpers.tf.read())
                helpers.mailout(helpers.MAILSUBJ_FP, output)
            sys.exit(1)

        msg = "Config found for CV's " + str(promote_list)
        helpers.log_msg(msg, 'DEBUG')

    # Get the org_id (Validates our connection to the API)
    org_id = helpers.get_org_id(org_name)

    # Now, let's fetch all available lifecycle environments for this org...
    (env_list, prior_list) = get_envs(org_id)

    # Get the list of Content Views along with the latest view version in each environment
    (ver_list, ver_descr, ver_version) = get_cv(org_id, target_env, env_list, prior_list,
        promote_list)

    # Promote to the given environment. Returns a list of task IDs.
    promote(target_env, ver_list, ver_descr, ver_version, env_list, prior_list, dry_run,
        args.quiet, args.forcemeta)

    # Add/Update the promotion history dictionary so we can check when we last promoted
    phistory[target_env] = datetime.datetime.strftime(datetime.datetime.now(), '%Y-%m-%d')
    pickle.dump(phistory, open(vardir + '/promotions.pkl', 'wb'))

    # Run the mailout
    if helpers.MAILOUT:
        helpers.tf.seek(0)
        output = "{}".format(helpers.tf.read())
        message = "Promotion completed successfully\n\n" + output
        subject = "Satellite 6 promotion completed"
        helpers.mailout(subject, message)

    # Exit cleanly
    sys.exit(0)

if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except KeyboardInterrupt, e:
        print >> sys.stderr, ("\n\nExiting on user cancel.")
        sys.exit(1)
